# Preparation of downloadable file from UKBRAP 
- Original files cannot be downloaded directly from UKBRAP 
1. CHR number and position per block of partitioned VCF (pVCF) file as listed in: <a href="https://biobank.ndph.ox.ac.uk/ukb/refer.cgi?id=837"></a>

2. Use SAK to filter out gene region: `bcftools view * -r chr6:46704201-46735721 > "$in-prefix"-filtered.vcf.gz`

3. Compress to download: `bcftools view -I * -O z -o compressed-filtered.vcf.gz`

# QC steps: 
## 1. From UKBB on-the-fly quality filtered exam (OQFE) protocol —  Hard QC

<a href="https://dnanexus.gitbook.io/uk-biobank-rap/science-corner/whole-exome-sequencing-oqfe-protocol/generation-and-utilization-of-quality-control-set-90pct10dp-on-oqfe-data/details-on-processing-the-300k-exome-data-to-generate-the-quality-control-set"></a> 

This protocol ensures variant and sample quality for exome sequencing data.

These coverage heterogeneities can also be mitigated by a single variant-level filter requiring that at least 90% of all genotypes for a given variant - independent of variant allele zygosity - have a read depth of at least 10 (i.e. DP>=10). When this filter is applied to the UKB WES 200k data prior to association analysis, the results are largely devoid of the spurious hits.

Variants are additionally filtered by genotype quality (<20), a genotype depth (<10), or >10% missing genotypes.

```
bcftools norm -m - -f GRCh38_full_analysis_set_plus_decoy_hla.fa -Oz -o normfiltered.vcf.gz <generegion.vcf.gz>

bcftools view -i 'F_PASS(DP>=10 | GT!="mis")> 0.9' -Oz -o hardQCnormfiltered.vcf.gz normfiltered.vcf.gz

bcftools view -i 'F_PASS(DP>=10 | GT!="mis" | GQ<20)> 0.9' -Oz -o hardQCnormfiltered.vcf.gz normfiltered.vcf.gz

bcftools filter -e '(FMT/GQ<20)' hardQCnormfiltered.vcf.gz -Oz -o GQfiltered.vcf.gz

```

Reference obtained from `wget https://ftp.ncbi.nlm.nih.gov/1000genomes/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa`

Remainder of QC from: <a href="https://doi.org/10.1038/s41586-021-04103-z"></a>

## 2. Variant-level QC

SNV genotypes with read depth (DP) less than 7 and indel genotypes with read depth less than 10 are changed to no-call genotypes. (Additional GQ>20) 

Note: the quotations need to be changed manually.

```
bcftools filter -S . -e '(TYPE="snp" & FMT/DP<7 & FMT/GQ<20 | TYPE="indel" & FMT/DP<10 & FMT/GQ<20)' hardQCnormfiltered.vcf.gz -Oz -o dpfiltered.vcf.gz

bcftools filter -e '(FMT/DP<10 | FMT/GQ<20 || GT!="mis">0.9)' hardQCnormfiltered.vcf.gz -Oz -o dpfiltered.vcf.gz
```
can use these filtering criteria: Requires ≥90% of genotypes per site have DP≥10 and not missing.
`'F_PASS(DP>=10 & GT!="mis")> 0.9'`

`-S .` changes to no call genotype, so number of variants remains the same as previous step (623).

After DP genotype filtering, a variant-level allele-balance filter is applied:

Retaining variants if:
1. At least one homozygous variant carrier, or
2. At least one heterozygous variant carrier with an allele balance (AB) greater than the cut-off (AB ≥ 0.15 for SNVs and AB ≥ 0.20 for indels).

AB for heterozygotes:
$AB = DP_[alt]/(DP_[ref]+DP_[alt])$

```
bcftools filter -i 'GT="hom" | (TYPE="snp" & GT="het" & (FORMAT/AD[1])/(FORMAT/AD+FORMAT/AD[1])>=0.15) | (TYPE="indel" & GT="het" & (FORMAT/AD[1])/(FORMAT/AD+FORMAT/AD[1])>=0.2)' dpfiltered.vcf.gz -Oz -o abdpfiltered.vcf.gz
  
bcftools filter -i 'GT=\"hom\" | TYPE=\"snp\" & GT=\"het\" & (FORMAT/AD[*:1])/(FORMAT/AD[*:0] + FORMAT/AD[*:1]) >= ${AB_snp} | TYPE=\"indel\" & GT=\"het\" & (FORMAT/AD[*:1])/(FORMAT/AD[*:0] + FORMAT/AD[*:1]) >= ${AB_indel}' -Oz -o "${_output}" 

```

Further SNVs are typically excluded if genotype missingness >5% or Hardy–Weinberg equilibrium test $p < 10^−6$ (GWAS convention, but rare variants were kept). 

* To achieve this, F_MISSING and HWE tag needs to be added for each variant site 
```
bcftools +fill-tags ABdphardqcnormfiltered.vcf.gz -Oz -z fillqcnormfitlered.vcf.gz -- -t F_MISSING,HWE 
```

* Filter based on added vcf tags
```
bcftools filter -I 'F_missing<0.05 & HWE>0.000001' -Oz -o finalfiltered-1.vcf.gz fillqcnormfiltered.vcf.gz 
```
Only removing HWE $p < 10^-6$
```
bcftools filter -I 'HWE>0.000001' -Oz -o finalfiltered-1.vcf.gz fillqcnormfiltered.vcf.gz 
```

## SAMPLE QC
Sample IDs were retrieved from cohort browser after filtering for each specific trait. 

Filtering of SNP IDs were done stepwise to identify number of individuals removed at each filtering stage. 

Converting comma separated ID text files to one sample per line: 
```
awk -F"," '{for(i=1;i<=NF;i++){print $i}}' <file name> > <editedfilename>
```

Remove space in front of line
```
sed 's/^[ \t]*//' <editfilename> > <finalfile>
```

### Exclusion criteria: 
1. Non-white European: Exclude ethnic background = white and IS NOT genetic ethnic grouping=caucasian
2. Genetically determined and reported sex:
- Two separate cohorts for male and female sex mismatch then combined
- Cohort 1: genetic sex is male and reported sex is female
- Cohort 2: genetic sex is female and reported sex is male

3. Heterozygosity/contamination:
- Outliers for heterozygosity or missing rate

4. Sex chromosome aneuploidy
- Sex chromosome aneuploid is YES

5. Consent: 
- Reason lost to follow up: participant has withdrawn consent for future linkage

6. Relatedness: 
- 10 or more third-degree relatives identified

Each filter was applied sequentially with bcftools: 
```
bcftools view -Oz -S ^<filenameofsnpstoexclude.txt> --force-samples <input vcf> > <output vcf>
```

# Formatting for rare variant tests

Files required:
1. Phenotype file (first 5 rows must match standard PLINK fields; use column 6 for phenotype)

2. Covariate file 
 
3. Genotype file (QCed sample and variant)
Chromosome should be number only (i.e. remove ‘chr’ prefix)

```
bcftools view <qced.vcf.gz> | awk '{gsub(/chr/, "")}1' > <newvcf>
bgzip -c <newvcf> > <zippedvcf> #zip 
tabix -f -p vcf <zippedvcf>  #index to tbi 
```

## FILE FORMAT: must have fiid, iid, fatid, matid, sex columns 
* Sex: female=0, male=1 

Changing female = 0, male = 1
```
awk '{if ($5=="Female") {$5=0} else {$5=1}}1' file.tsv > fileedited.tsv
awk '{if ($5=="Female") {$5=0} else {$5=1}}1' covariates_participant.tsv >covarNA.tsv
```

Changing header back to sex
```
sed -e '1s/"1"/Sex/' <input> > <output>
```

At this stage, all data is prepped for rare variant association analyses

# Formatting for linear regression in R

Extract sample IDs from QCed vcf file – to obtain EIDs for those with genetic data only 
```
bcftools query -l <qced file> > <sampleid.txt>
```

Subsetting phenofile to match sampleids
```
grep -wFf <sampleid.txt> <phenofile.tsv> > <phenowgenoid.tsv>
```

Subset to carrier phenotypes 
```
grep -wFf <carrierID.txt> < phenowgenoid.tsv> > <carrierpheno.tsv>
```

Subset control phenotypes
```
grep -v -wFf <carrierID.txt> < phenowgenoid.tsv> > <carrierpheno.tsv>
## tip: -v option in grep inverts matching 
```